### PR TITLE
feat: Add CompletableFuture WebStorage API

### LINKED ISSUE #19743: Improve WebStorage API
### Describe your motivation

I'm programming in kotlin and heavily using coroutines. The WebStorage API only defines static functions with callbacks, which can only be extended as suspending in Kotlin with:

```kotlin
suspend fun getWebstorageItem(key: String): String? {
    return suspendCoroutine { continuation ->
        WebStorage.getItem(key) { continuation.resume(it) }
    }
}
```

Because these are static methods I couldn't do:

```kotlin
suspend fun WebStorage.getItem(key: String): String? {
    return suspendCoroutine { continuation ->
        getItem(key) { continuation.resume(it) }
    }
}
```

### Describe the solution you'd like

That's why I suggest to add overload methods that return a CompletableFuture instead of taking a callback. This way I can easily await on the CompletableFuture. Like:

```java
    public static CompletableFuture getItem(String key) {
        // Impl Detail
    }
```

```kotlin
     WebStorage.getItem("key").await()
```

This way also other reactive frameworks would benefit from it.
The API could be expanded like this without breaking the existing one.

### Describe alternatives you've considered

Also I would suggest to refactor the API to use objects. So that the calls could be:

```java
   WebStorage.getLocalStorage().setItem("key")
   WebStorage.getSessionStorage().setItem("key")
   
   WebStorage.getLocalStorage().getItem("key")
   WebStorage.getSessionStorage().getItem("key")
```

For spsecific UI:

```java   
   UI ui = UI.getCurrent()
   WebStorage.getLocalStorage(ui).setItem("key")
   WebStorage.getSessionStorage(ui).setItem("key")
   
   WebStorage.getLocalStorage(ui).getItem("key")
   WebStorage.getSessionStorage(ui).getItem("key")
```
Or as alternative construct it directly as object:

```java   
   new LocalStorage(UI.getCurrent()).getItem("key")
   new SessionStorage(UI.getCurrent()).getItem("key")   
```

All my suggestions would allow to keep the current static approach without breaking.

### PR DESCRIPTION
**Pull request by Dudeplayz**
## Description

Adds CompletableFuture methods for `WebStorage.getItem`.

Fixes #19743

## Type of change

- [ ] Bugfix
- [x] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/guide/contributing/overview/
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [ ] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [x] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

**Comment by Dudeplayz** [comment]
Questions:
- The callback style methods could also use the CompletableFuture one and be called with then?
- CompletableFuture would return an exception in error cases instead of null, how is the behavior intended here?

**Comment by mcollovati** [comment]
> The callback style methods could also use the CompletableFuture one and be called with then?

I would leave the callback method as is. I think there's no need to create a `CompletableFuture` there.

> CompletableFuture would return an exception in error cases instead of null, how is the behavior intended here?

My opinion is that the `CompletableFuture` version should behave in the same way as the `Callback` one. So, probably the returned `CompletableFuture` should have an `exceptionally(...)` stage that returns null (and maybe logging the error at DEBUG level). It would also be good to mention the behavior on the Javadoc.

File: flow-server/src/main/java/com/vaadin/flow/component/page/WebStorage.java [hunk 1]
```diff
@@ -215,17 +219,88 @@ public static void getItem(Storage storage, String key, Callback callback) {
...
+     * while holding the session lock since the request handling thread that
+     * makes the result available will also need to lock the session. <br>
+     * See {@link PendingJavaScriptResult#toCompletableFuture} for more
+     * information.
+     *
+     * @param key
+     *            the key for which the value will be fetched
+     * @return a CompletableFuture that will be completed with the value once
+     *         transferred from the client side or <code>null</code> if the
+     *         value was not set
```
Comment (mcollovati) [inline flow-server/src/main/java/com/vaadin/flow/component/page/WebStorage.java:249]: I would use `available` as in the `Callback` javadoc
```suggestion
     *         value was not available.
```

File: flow-server/src/main/java/com/vaadin/flow/component/page/WebStorage.java [hunk 2]
```diff
@@ -215,17 +219,88 @@ public static void getItem(Storage storage, String key, Callback callback) {
...
+     * See {@link PendingJavaScriptResult#toCompletableFuture} for more
+     * information.
+     *
+     * @param storage
+     *            the storage
+     * @param key
+     *            the key for which the value will be fetched
+     * @return a CompletableFuture that will be completed with the value once
+     *         transferred from the client side or <code>null</code> if the
+     *         value was not set
```
Comment (mcollovati) [inline flow-server/src/main/java/com/vaadin/flow/component/page/WebStorage.java:269]:
```suggestion
     *         value was not available.
```

File: flow-server/src/main/java/com/vaadin/flow/component/page/WebStorage.java [hunk 3]
```diff
@@ -215,17 +219,88 @@ public static void getItem(Storage storage, String key, Callback callback) {
...
+     *
+     * @param ui
+     *            the UI for which the storage is related to
+     * @param storage
+     *            the storage
+     * @param key
+     *            the key for which the value will be fetched
+     * @return a CompletableFuture that will be completed with the value once
+     *         transferred from the client side or <code>null</code> if the
+     *         value was not set
```
Comment (mcollovati) [inline flow-server/src/main/java/com/vaadin/flow/component/page/WebStorage.java:293]:
```suggestion
     *         value was not available.
```

**Review by mcollovati — CHANGES_REQUESTED** [review]

**Review by mcollovati — APPROVED** [review]

**Pull request merged**